src/go/k8s/internal/client/spec_sasl.go

```diff
@@ -0,0 +1,164 @@
+package client
+
+import (
+	"context"
+	"fmt"
+
+	krbclient "github.com/jcmturner/gokrb5/v8/client"
+	krbconfig "github.com/jcmturner/gokrb5/v8/config"
+	"github.com/jcmturner/gokrb5/v8/keytab"
+	"github.com/redpanda-data/console/backend/pkg/config"
+	redpandav1alpha2 "github.com/redpanda-data/redpanda-operator/src/go/k8s/api/redpanda/v1alpha2"
+	"github.com/twmb/franz-go/pkg/kgo"
+	"github.com/twmb/franz-go/pkg/sasl"
+	"github.com/twmb/franz-go/pkg/sasl/aws"
+	"github.com/twmb/franz-go/pkg/sasl/kerberos"
+	"github.com/twmb/franz-go/pkg/sasl/oauth"
+	"github.com/twmb/franz-go/pkg/sasl/plain"
+	"github.com/twmb/franz-go/pkg/sasl/scram"
+)
+
+func (c *clientFactory) configureAdminSpecSASL(ctx context.Context, namespace string, spec *redpandav1alpha2.AdminAPISpec) (username string, password string, token string, err error) {
+	if spec.SASL == nil {
+		return "", "", "", nil
+	}
+
+	switch spec.SASL.Mechanism {
+	// SCRAM
+	case config.SASLMechanismScramSHA256, config.SASLMechanismScramSHA512:
+		p, err := spec.SASL.Password.GetValue(ctx, c.Client, namespace, "password")
+		if err != nil {
+			return "", "", "", fmt.Errorf("unable to fetch sasl password: %w", err)
+		}
+
+		return spec.SASL.Username, string(p), "", nil
+	// OAUTH
+	case config.SASLMechanismOAuthBearer:
+		token, err := spec.SASL.AuthToken.GetValue(ctx, c.Client, namespace, "password")
+		if err != nil {
+			return "", "", "", fmt.Errorf("unable to fetch sasl token: %w", err)
+		}
+		return "", "", string(token), nil
+	}
+
+	return "", "", "", fmt.Errorf("unsupported SASL mechanism: %s", spec.SASL.Mechanism)
+}
+
+func (c *clientFactory) configureKafkaSpecSASL(ctx context.Context, namespace string, spec *redpandav1alpha2.KafkaAPISpec) (kgo.Opt, error) {
```

Comment:
This is split up from the admin-based SASL code above since we don't support the same subsets of authentication in our SASL implementations across APIs.